tech-spec.md

```diff
@@ -0,0 +1,310 @@
+# Tech spec
+
+The ODM sharing library aims to be able to read the most used file formats and
+databases, provide useful feedback to help build correct filters, and output
+easy-to-share result data.
+
+## Overview
+
+A sharing schema (in CSV format) is used to define individual rules (as rows),
+which come together to form a data query.
+
+The query rules are parsed into an abstract syntax tree (AST), which represents
+the structure of the query. It effectively separates rule-parsing from
+query-generation, making it more modular. The AST can then be used to generate
+concrete queries in SQL or other query languages, or it can be interpreted to
+perform data transformations directly using a library like Pandas.
+
+The library will focus on using SQL as the query language, since it has wide
+support and is easy to generate. Using SQL will also provide a free performance
+boost when running queries on big indexed databases, as well as enable us to
+output the intermediate SQL to the user in case they want to inspect or execute
+the queries themselves.
+
+## CLI
+
+Usage:
+
+```bash
+./share.py [--org=<name>] [--outdir=<path>] <schema> <input>
+```
+
+- org: the organization to output data for, or all if not specified
+- outdir: output file directory
+- schema: sharing schema file path
+- input: spreadsheet file path or [SQLAlchemy database url](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls)
+Examples:
+
+```bash
+./share.py --org=OHRI --outdir=~/ohri/ schema.csv data.xlsx
+./share.py schema.csv mysql+pymysql://scott:tiger@localhost/foo
+./share.py schema.csv mssql+pyodbc://user:pass@mydsn
+```
+
+## API
+
+### Public modules
+
+#### sharing
+
+- high level:
+    ```python
+    extract(data_source: str, schema_file: str, org: str=''
+            ) -> Dict[TableName, Dataframe]
+    ```
+    - data_source: a file path or database url (in SQLAlchemy format)
+    - schema_file: rule schema file path
+    - org: org to share with, or all orgs if empty
+    - returns a Pandas Dataframe per table, for further processing
```

Comment:
For the file, adding the organization in the file name makes sense. But for the function I assume that it will be an object with two levels of nesting? The first level contains the organization, the next level contains the tables, and the last level has the data?

Reply:
you're right, i forgot to add the org as part of the `extract` return type, like the one `parse` has. I'll fix that.